### PR TITLE
Implement #13384: Expose all TileElement data to plugin API

### PR DESCRIPTION
**Pull request by Sadret**
#### Adds the following fields:
TileElement.isGhost
TileElement.baseZ (= baseHeight * 8)
TileElement.clearanceZ (= clearanceHeight * 8)
Footpath.edges
Footpath.corners
Track.mazeEntry
Track.colourScheme
Track.seatRotation
Track.brakeBoosterSpeed ( = stored value * 2)
Track.isInverted
Track.hasCableLift
Wall.slope
Wall.primaryColour
Wall.secondaryColour
Wall.tertiaryColour
Wall.bannerIndex
LargeScenery.bannerIndex
LargeScenery.sequence
Banner.bannerIndex
Banner.direction

#### Removes the following fields:
Footpath.edgesAndCorners (replaced by Footpath.edges and Footpath.corners)
Footpath.railings (unused)
Footpath.direction (unused)

#### Changes to openrct2.d.ts
- adds/removes above fields
- removes FootpathElement->footpathType (unused / mistake?)
- adds nullable types where needed
- removes type TileElement (wrongly used and superfluous / mistake?)

#### Other changes:
- API returns null values for all fields that do not make sense for that element, e.g. corners for Walls.
- Banner.position is merged into TileElement.direction, as they have the same semantic meaning. (Note that above this change only affect the plugin API and nothing else.)
- Increased API version.

### CODE REVIEW

File: src/openrct2/scripting/ScTile.hpp [hunk 1]
```diff
@@ -112,6 +112,17 @@ namespace OpenRCT2::Scripting
             Invalidate();
         }
 
+        uint8_t baseZ_get() const
+        {
+            return _element->GetBaseZ();
+        }
+        void baseZ_set(uint8_t value)
```
Comment (Gymnasiast) [inline src/openrct2/scripting/ScTile.hpp:119]: Should be an uin16_t at least. Valid range is `0 - (254 * COORDS_Z_STEP)`. (Because 255 is used to mark the tile element as discarded.

Comment (Sadret) [reply src/openrct2/scripting/ScTile.hpp]: Same for clearanceZ.
Changed it to uint16_t.

File: src/openrct2/scripting/ScTile.hpp [hunk 2]
```diff
@@ -719,37 +910,113 @@ namespace OpenRCT2::Scripting
...
         }
 
-        bool railings_get() const
+        uint8_t edges_get() const
         {
             auto el = _element->AsPath();
-            return el != nullptr ? el->GetRailingEntryIndex() : false;
+            return el != nullptr ? el->GetEdges() : 0;
         }
-        void railings_set(bool value)
```
Comment (Gymnasiast) [inline src/openrct2/scripting/ScTile.hpp:730]: Why did you remove railings_get / railings_set ?

Comment (Sadret) [reply src/openrct2/scripting/ScTile.hpp:730]: It's unused. The getter is just an alias for `GetSurfaceEntryIndex` and the setter does nothing.

Comment (Gymnasiast) [reply src/openrct2/scripting/ScTile.hpp:730]: .. for now. But well, how path surfaces and railings work is bound to change a lot anyway.

**Review by Gymnasiast — APPROVED** [review]
LGTM, but would like either @IntelOrca or @duncanspumpkin to review it as well.

File: src/openrct2/scripting/ScTile.hpp [hunk 3]
```diff
@@ -446,6 +501,114 @@ namespace OpenRCT2::Scripting
             }
         }
 
+        uint16_t mazeEntry_get() const
+        {
+            auto el = _element->AsTrack();
+            if (el != nullptr)
```
Comment (duncanspumpkin) [inline src/openrct2/scripting/ScTile.hpp:507]: Should we have an extra check to make sure this is a maze entry?

File: src/openrct2/scripting/ScTile.hpp [hunk 4]
```diff
@@ -446,6 +501,114 @@ namespace OpenRCT2::Scripting
...
+                return el->GetMazeEntry();
+            return 0;
+        }
+        void mazeEntry_set(uint16_t value)
+        {
+            ThrowIfGameStateNotMutable();
+            auto el = _element->AsTrack();
+            if (el != nullptr)
```
Comment (duncanspumpkin) [inline src/openrct2/scripting/ScTile.hpp:515]: As above

File: src/openrct2/scripting/ScTile.hpp (hunk 3 above)
Comment (Sadret) [reply src/openrct2/scripting/ScTile.hpp:507]: Probably. Consequently, we should then also check in `sequence_get` that it is _not_ a maze. Also:
`Footpath->AdditionStatus` vs `Footpath->rideIndex`
`Track->ColourScheme` (i.e. `colourScheme` and `seatRotation`) vs `Track->MazeEntry`
`Track->BrakeBoosterSpeed` vs it's not a break/booster.
That's all unions used.

Comment (duncanspumpkin) [reply src/openrct2/scripting/ScTile.hpp:507]: @IntelOrca what are your opinions

Comment (IntelOrca) [reply src/openrct2/scripting/ScTile.hpp:507]: I would say so.

File: distribution/openrct2.d.ts [hunk 5]
```diff
@@ -531,66 +534,88 @@ declare global {
...
+    interface WallElement extends BaseTileElement {
+        direction: Direction;
         object: number;
-        sequence: number;
-        ride: number;
-        station: number;
+        primaryColour: number;
+        secondaryColour: number;
+        tertiaryColour: number;
+        bannerIndex: number;
```
Comment (duncanspumpkin) [inline distribution/openrct2.d.ts:589]: bannerIndex could be null

File: distribution/openrct2.d.ts [hunk 6]
```diff
@@ -531,66 +534,88 @@ declare global {
...
+        sequence: number;
+        footpathObject: number;
     }
 
     interface LargeSceneryElement extends BaseTileElement {
+        direction: Direction;
         object: number;
         primaryColour: number;
         secondaryColour: number;
+        bannerIndex: number;
```
Comment (duncanspumpkin) [inline distribution/openrct2.d.ts:607]: bannerIndex could be null

File: distribution/openrct2.d.ts (hunk 5 above)
Comment (Sadret) [reply distribution/openrct2.d.ts:589]: The way it is implemented right now, bannerIndex is never null. But bannerIndex == 0 is ambigious, as it can either mean no banner or the first banner.
I can change that by e.g. checking if GetBanner() == nullptr first.

File: distribution/openrct2.d.ts (hunk 6 above)
Comment (Sadret) [reply distribution/openrct2.d.ts:607]: see above

**Review by duncanspumpkin — APPROVED** [review]
Don't forget to increment the api version

**Comment by Sadret** [comment]
> Don't forget to increment the api version

I did, twice. Now a third time, I guess.

**Comment by Sadret** [comment]
Changed
`tile_element_get_banner_index(_element)`
to
`_element->GetBannerIndex();`
after the latest rebase.

**Comment by duncanspumpkin** [comment]
@Sadret for future what I do is wait until all of the reviews are in and no further changes are being asked and then bump the network/plugin version. Means you don't have to do it as many times. Doesn't always work out like that though :(.
